### PR TITLE
Reject pointless select/group by AID

### LINKED ISSUE #397: Reject queries directly selecting AID columns.
Such queries will take a long time to finish and have no chance of returning any useful results.

### PR DESCRIPTION
**Pull request by pdobacz**
Closes #397

### CODE REVIEW

File: src/query/anonymization.c [hunk 1]
```diff
@@ -471,6 +493,8 @@ static AnonymizationContext *make_query_anonymizing(Query *query, List *personal
     anon_context->expand_buckets = true;
   }
 
+  reject_aid_grouping(query, aid_refs);
```
Comment (pdobacz) [inline src/query/anonymization.c:496]: I pondered putting this under `validation.c`, however, firstly selecting AIDs doesn't make the query invalid in any sense, only pointless.

Secondly, under `validation.c` we'd need to have `personal_relations` handy, and I don't want to include it there without a strong reason.

File: test/expected/validation.out [hunk 2]
```diff
@@ -179,25 +179,25 @@ FROM (
      0
 (1 row)
 
-SELECT COUNT(*), COUNT(x.city), COUNT(DISTINCT x.id)
+SELECT COUNT(*), COUNT(x.city), COUNT(DISTINCT x.city)
```
Comment (pdobacz) [inline test/expected/validation.out:182]: I'm not exactly sure, how much these 3 tests were specifically about operations on an AID column. If they don't make sense for non-aid column, please let me know

File: src/query/anonymization.c [hunk 3]
```diff
@@ -270,6 +270,28 @@ static List *gather_aid_refs(Query *query, List *relations)
...
+    if (IsA(group_expr, Var))
+    {
+      Var *var = (Var *)group_expr;
+
+      ListCell *aid_ref_cell;
+      foreach (aid_ref_cell, aid_refs)
+      {
+        AidRef *aid_ref = (AidRef *)lfirst(aid_ref_cell);
+        if (aid_ref->aid_attnum == var->varattno)
+          FAILWITH_LOCATION(var->location, "Selecting AID without generalization cannot yield any results - rejecting.");
```
Comment (pdobacz) [inline src/query/anonymization.c:289]: I picked the word "Selecting" here, but the problem lies actually in grouping/bucket expression. However, I'm not sure mentioning that is the most useful error message in this context. Anyone has ideas/thoughts on this?

File: src/query/anonymization.c [hunk 4]
```diff
@@ -270,6 +270,28 @@ static List *gather_aid_refs(Query *query, List *relations)
...
+    Node *group_expr = (Node *)lfirst(cell);
+    if (IsA(group_expr, Var))
+    {
+      Var *var = (Var *)group_expr;
+
+      ListCell *aid_ref_cell;
+      foreach (aid_ref_cell, aid_refs)
+      {
+        AidRef *aid_ref = (AidRef *)lfirst(aid_ref_cell);
+        if (aid_ref->aid_attnum == var->varattno)
```
Comment (cristianberneanu) [inline src/query/anonymization.c:288]: The relation id also needs to be matched here.

File: src/query/anonymization.c (hunk 1 above)
Comment (cristianberneanu) [reply src/query/anonymization.c:496]: You don't need `personal_relations` to check for AID columns, look at `is_aid_column`, which reads the security label directly. I think this should be called by either `verify_anonymization_requirements` or by `compile_anonymizing_query`.

File: test/expected/validation.out (hunk 2 above)
Comment (cristianberneanu) [reply test/expected/validation.out:182]: Why is this rejected? Aggregating AID columns should be allowed. Same thing below.

Comment (pdobacz) [reply test/expected/validation.out:182]: it SELECTs * from a personal relation, therefore including `id` - the AID

Comment (cristianberneanu) [reply test/expected/validation.out:182]: You're right, stupid question on my part... I completely missed the `*` select in the subquery while looking at it.

File: src/query/anonymization.c [hunk 5]
```diff
@@ -270,24 +271,21 @@ static List *gather_aid_refs(Query *query, List *relations)
...
+      RangeTblEntry *rte = rt_fetch(var->varno, query->rtable);
 
-      ListCell *aid_ref_cell;
-      foreach (aid_ref_cell, aid_refs)
-      {
-        AidRef *aid_ref = (AidRef *)lfirst(aid_ref_cell);
-        if (aid_ref->aid_attnum == var->varattno)
-          FAILWITH_LOCATION(var->location, "Selecting AID without generalization cannot yield any results - rejecting.");
-      }
+      if (rte->relkind == 'r' && is_aid_column(rte->relid, var->varattno))
```
Comment (edongashi) [inline src/query/anonymization.c:287]: There is a `RELKIND_RELATION` define.

Comment (pdobacz) [reply src/query/anonymization.c]: ok, force pushed

**Review by cristianberneanu — APPROVED** [review]

**Pull request merged**